### PR TITLE
[1.17]further simplification of overwrite_specials attribute code

### PR DESCRIPTION
**Pull request by newfrenchy83**
I know that I have already proposed a simplification of the code of this attribute, but several things lead me to propose this second rewrite.

First, using the first loop to define three boolean variables is more complicated than simply eliminating from the 'overwrite_specials' list the abilities not carrying the attribute then; if an ability with the attribute is present define a double for loop inside each other, where for each ability with overwrite_specials the type of ability removal from the returned list is defined at that time (one_side or both_sides). The code is even more readable than with the previous work in my opinion.

The second reason is that I eventually want to be able to use the request code https://github.com/wesnoth/wesnoth/pull/7050 to be even more selective about the abilities that can be removed or not and extend the use of the attribute beyond chance_to_hit (when (and if) these two PRs will be merged in master, I will have to add 3-4 lines of code in hadcoding to implement the filtering). Building a list instead of splicing the existing one is incompatible with this project.

### CODE REVIEW

File: src/units/attack_type.hpp
Comment (stevecotton) [inline src/units/attack_type.hpp:137]: I found the documentation a bit confusing, and think this would be clearer (including renaming some of the arguments). Please would you check if it's correct?

Assuming it is correct, you can either cherry-pick it or I can create a PR for it afterwards.

```
	/**
	 * Filter a list of abilities or weapon specials, removing any entries that are overridden by
	 * the overwrite_specials attributes of a second list.
	 *
	 * @param ability The special ability type who is being checked.
	 * @param input list to check, a filtered copy of this list is returned by the function.
	 * @param overwriters list that may have overwrite_specials attributes.
	 * @param filter_self name of [filter_"self/student"] if is abilities or specials who are checked.
	 * @param is_special if true, input contains weapon specials; if false, it contains abilities.
	 */
	unit_ability_list overwrite_special_checking(const std::string& ability, unit_ability_list input, unit_ability_list overwriters, const std::string& filter_self, bool is_special) const;
```

File: src/units/abilities.cpp
Comment (stevecotton) [inline src/units/abilities.cpp:1317]: If `temp_overwrite` is false now, then all assignments to `is_overwritable` will also be false, because the ones below start with `temp_overwrite &&`.

**Review by stevecotton — CHANGES_REQUESTED** [review]
Thanks - it's much easier to review a small refactor. I have a few comments, but my general feeling is that this will merge soon.

File: src/units/abilities.cpp
Comment (newfrenchy83) [reply src/units/abilities.cpp:1317]: is_overwritable can't be true if is an ability with overwrite_specials attribute is_checked, then yes i will rephrase the code of if loop.

File: src/units/abilities.cpp
Comment (newfrenchy83) [reply src/units/abilities.cpp]: for one function only, idon't know.

Comment (stevecotton) [reply src/units/abilities.cpp]: I'm looking at doing that refactor elsewhere in the codebase, will open a followup PR after this one merges.

File: src/units/attack_type.hpp
Comment (stevecotton) [inline src/units/attack_type.hpp:130]: Trivial issue - the whitespace is wrong on this line.

File: src/units/abilities.cpp
Comment (stevecotton) [inline src/units/abilities.cpp:1327]: The line above could be combined with this one, so it's just `if(is_overwritable && one_side_overwritable)`. However, no need to do that if your next PR is going to add additional logic that's easier with them kept separate.

**Review by stevecotton — APPROVED** [review]
@CelticMinstrel this looks good to me, but will wait until you've had time to look too.

File: src/units/attack_type.hpp
Comment (newfrenchy83) [reply src/units/attack_type.hpp]: corrected

**Comment by CelticMinstrel** [comment]
I'm afraid I have no idea what this is doing, so I don't think my review would be helpful…

**Comment by newfrenchy83** [comment]
> I'm afraid I have no idea what this is doing, so I don't think my review would be helpful…

This is another code simplification of the overwrite_specials attribute to pave the way for adding a filter that will use code from https://github.com/wesnoth/wesnoth/pull/7050.
Otherwise, the operation of the attribute will not be affected and the code will be even easier to understand than before.

**Comment by stevecotton** [comment]
My plan is to merge soon, just waiting for the CI to confirm that the build error introduced in a7b39bd8996a847fd4c2160bdb6d12b3efdc67f5 has been fixed.

**Comment by newfrenchy83** [comment]
> My plan is to merge soon, just waiting for the CI to confirm that the build error introduced in [a7b39bd](https://github.com/wesnoth/wesnoth/commit/a7b39bd8996a847fd4c2160bdb6d12b3efdc67f5) has been fixed.

it seem what yes.

**Pull request merged**
